File: src/mutateAndScore/mutator.py
import random
from enum import Enum

from lib.rgb import rgb
from lib.svg import SVG, Circle, Coloring, Outline, Polygon, Rectangle
from lib.svg.elements.utils.mathHelpers import randint

from ..randomize.randomSVG import getRandShape


class Mutation(Enum):
    GEOMETRY = 0.40
    APPEARANCE = 0.30
    SWAP_LAYER = 0.15
    ADD = 0.10
    REMOVE = 0.05


def choose_mutation():
    r = random.random()

    if r < 0.40:
        return 0
    if r < 0.70:
        return 1
    if r < 0.85:
        return 2
    if r < 0.95:
        return 3
    return 4


class Mutator:
    __slots__ = ("elms", "height", "svg", "width")

    def __init__(self, svg: SVG, height, width):
        self.svg = svg
        self.elms = svg.data
        self.height = height
        self.width = width

        self.mutate()

    def get(self):
        self.svg.setData(self.elms)
        return self.svg

    def mutate(self):
        elms = self.elms

        if not elms:
            self.addShape()
            return

        strategy = choose_mutation()
        n = len(elms)

        if strategy == 0:
            self.alterGeometry()
        elif strategy == 1:
            self.alterAppearance()
        elif strategy == 2:
            if n >= 2:
                self.swapLayer()
            else:
                self.alterGeometry()
        elif strategy == 3:
            self.addShape()
        elif n > 1:
            self.removeShape()
        else:
            self.alterGeometry()

    def swapLayer(self):
        elms = self.elms
        n = len(elms)

        i = random.randrange(n)
        j = random.randrange(n - 1)

        if j >= i:
            j += 1

        elms[i], elms[j] = elms[j], elms[i]

    def alterGeometry(self):
        elm = random.choice(self.elms)

        dx, dy = random.randint(-5, 5), random.randint(-5, 5)
        if isinstance(elm, Polygon):
            positions = elm.positions
            i = random.randrange(len(positions))
            x, y = positions[i]

            elm.updatePoint((x + dx, y + dy), i)

        elif isinstance(elm, Circle):
            elm.changeCenter(elm.cx + dx, elm.cy + dy)

        elif isinstance(elm, Rectangle):
            elm.changeTopLeftCorner(elm.x + dx, elm.y + dy)

    def addShape(self):
        self.elms.append(getRandShape().generate(self.height, self.width))

    def alterAppearance(self):
        elm = random.choice(self.elms)

        r, g, b = randint(255, len=3)

        if random.getrandbits(1):
            inner = elm.inner
            baseOpacity = inner.opacity if inner is not None else 0

            opacity = baseOpacity + random.uniform(-0.1, 0.1)
            opacity = max(0.05, min(1.0, opacity))

            elm.updateColoring(Coloring(fill=rgb(r, g, b), opacity=round(opacity, 2)))

        else:
            outer = elm.outer

            if outer is None:
                baseOpacity = 0
                baseWidth = 0
            else:
                baseOpacity = outer.opacity
                baseWidth = outer.width

            opacity = baseOpacity + random.uniform(-0.1, 0.1)
            opacity = max(0.05, min(1.0, opacity))

            width = max(0.5, baseWidth + random.uniform(-0.5, 0.5))

            elm.updateOutline(
                Outline(
                    stroke=rgb(r, g, b),
                    width=round(width, 2),
                    opacity=round(opacity, 2),
                )
            )

    def removeShape(self):
        self.elms.pop(random.randrange(len(self.elms)))

File: src/mutateAndScore/mutator.cc
#include "mutator.hh"

#include <algorithm>
#include <array>
#include <cmath>
#include <optional>
#include <random>

#include "../../lib/svg/svg.hh"
#include "../../lib/svg/elements/appearance.hh"
#include "../randomize/randomSVG.hh"


namespace {

std::mt19937& rng() {
    static thread_local std::mt19937 engine{std::random_device{}()};
    return engine;
} 

double round2(double v) { return std::round(v * 100.0) / 100.0; }

double opacityOr(const std::optional<Coloring>& c, double fallback) {
    return c.has_value() ? c->opacity : fallback;
}

double opacityOr(const std::optional<Outline>& o, double fallback) {
    return o.has_value() ? o->opacity : fallback;
}

double widthOr(const std::optional<Outline>& o, double fallback) {
    return o.has_value() ? o->width : fallback;
}

constexpr std::array<double, 5> kWeights = {0.40, 0.30, 0.15, 0.10, 0.05};

} // anonymous

Mutator::Mutator(SVG& svg, int height, int width)
    : svg_(svg), height_(height), width_(width) {
    mutate();
}

SVG& Mutator::get() {
    svg_.setData(svg_.data());
    return svg_;
}

void Mutator::mutate() {
    auto& elms = svg_.data();
    if (elms.empty()) {
        addShape();
        return;
    }

    std::discrete_distribution<int> dist(kWeights.begin(), kWeights.end());
    const auto strategy = static_cast<MutationKind>(dist(rng()));

    if (strategy == MutationKind::SwapLayer && elms.size() >= 2) {
        swapLayer();
    } else if (strategy == MutationKind::Geometry) {
        alterGeometry();
    } else if (strategy == MutationKind::Appearance) {
        alterAppearance();
    } else if (strategy == MutationKind::Add) {
        addShape();
    } else if (strategy == MutationKind::Remove && elms.size() > 1) {
        removeShape();
    } else {
        alterGeometry();
    }
}

void Mutator::swapLayer() {
    auto& elms = svg_.data();
    std::uniform_int_distribution<size_t> dist(0, elms.size() - 1);

    size_t idx1 = dist(rng());
    size_t idx2 = dist(rng());
    while (idx2 == idx1) idx2 = dist(rng());

    std::swap(elms[idx1], elms[idx2]);
}

void Mutator::alterGeometry() {
    auto& elms = svg_.data();
    std::uniform_int_distribution<size_t> pick(0, elms.size() - 1);
    auto& elm = elms[pick(rng())];

    std::uniform_int_distribution<int> jitter(-5, 5);

    if (auto* poly = dynamic_cast<Polygon*>(elm.get())) {
        std::uniform_int_distribution<size_t> ptDist(0, poly->positions().size() - 1);
        const size_t idx = ptDist(rng());
        const Point p = poly->positions()[idx];
        poly->updatePoint({p.x + jitter(rng()), p.y + jitter(rng())}, idx);

    } else if (auto* ell = dynamic_cast<Ellipse*>(elm.get())) {
        ell->changeCenter(ell->x + jitter(rng()), ell->y + jitter(rng()));

    } else if (auto* circ = dynamic_cast<Circle*>(elm.get())) {
        circ->changeCenter(circ->x + jitter(rng()), circ->y + jitter(rng()));

    } else if (auto* rect = dynamic_cast<Rectangle*>(elm.get())) {
        rect->changeTopLeftCorner(rect->x + jitter(rng()), rect->y + jitter(rng()));
    }
}

void Mutator::addShape() {
    auto factory = random_svg::getRandShapeFactory();
    svg_.data().push_back(factory(height_, width_));
}

void Mutator::alterAppearance() {
    auto& elms = svg_.data();
    std::uniform_int_distribution<size_t> pick(0, elms.size() - 1);
    auto& elm = elms[pick(rng())];

    std::uniform_int_distribution<int> colorDist(0, 255);
    const int r = colorDist(rng()), g = colorDist(rng()), b = colorDist(rng());

    std::bernoulli_distribution coin(0.5);
    std::uniform_real_distribution<double> deltaOpacity(-0.1, 0.1);

    if (coin(rng())) { // Coloring
        const double baseOpacity = opacityOr(elm->inner(), 0.0);
        const double opacity = std::clamp(baseOpacity + deltaOpacity(rng()), 0.05, 1.0);
        elm->updateColoring(Coloring(rgb(r, g, b), round2(opacity)));

    } else { // Outline
        const double baseOpacity = opacityOr(elm->outer(), 0.0);
        const double baseWidth = widthOr(elm->outer(), 0.0);

        std::uniform_real_distribution<double> deltaWidth(-0.5, 0.5);
        const double opacity = std::clamp(baseOpacity + deltaOpacity(rng()), 0.05, 1.0);
        const double width = std::max(0.5, baseWidth + deltaWidth(rng()));

        elm->updateOutline(Outline(rgb(r, g, b), round2(width), round2(opacity)));
    }
}

void Mutator::removeShape() {
    auto& elms = svg_.data();
    std::uniform_int_distribution<size_t> dist(0, elms.size() - 1);
    elms.erase(elms.begin() + dist(rng()));
}